client/src/main/java/com/alibaba/nacos/client/naming/remote/gprc/NamingGrpcConnectionEventListener.java

```diff
@@ -46,48 +48,81 @@
     
     private final Set<String> subscribes = new ConcurrentHashSet<String>();
     
+    private volatile boolean connected = false;
+    
+    private static final long DEFAULT_REDO_DELAY = 3000L;
+    
+    private static final int DEFAULT_REDO_THREAD = 2;
+    
+    private ScheduledExecutorService redoExecutorService;
+    
     public NamingGrpcConnectionEventListener(NamingGrpcClientProxy clientProxy) {
         this.clientProxy = clientProxy;
+        this.redoExecutorService = new ScheduledThreadPoolExecutor(DEFAULT_REDO_THREAD, r -> {
+            Thread t = new Thread(r);
+            t.setName("com.alibaba.nacos.client.naming.grpc.event.listener");
+            t.setDaemon(true);
+            return t;
+        });
     }
     
     @Override
     public void onConnected() {
-        redoSubscribe();
-        redoRegisterEachService();
+        connected = true;
+        LogUtils.NAMING_LOGGER.info("Grpc re-connect, redo subscribe services");
+        redoSubscribe(subscribes);
+        LogUtils.NAMING_LOGGER.info("Grpc re-connect, redo register services");
+        redoRegisterEachService(registeredInstanceCached.keySet());
     }
     
-    private void redoSubscribe() {
-        LogUtils.NAMING_LOGGER.info("Grpc re-connect, redo subscribe services");
+    private void redoSubscribe(Set<String> subscribes) {
+        Set<String> failedSubscribes = new ConcurrentHashSet<>();
         for (String each : subscribes) {
             ServiceInfo serviceInfo = ServiceInfo.fromKey(each);
             try {
                 clientProxy.subscribe(serviceInfo.getName(), serviceInfo.getGroupName(), serviceInfo.getClusters());
             } catch (NacosException e) {
-                LogUtils.NAMING_LOGGER.warn(String.format("re subscribe service %s failed", serviceInfo.getName()), e);
+                if (connected) {
```

Comment:
Can we move the connected judgement in start of loop?

Reply:
OK,  already moved.